packages/dds/tree/src/simple-tree/toMapTree.ts

```diff
@@ -353,26 +330,51 @@ function checkInput(condition: boolean, message: string | (() => string)): asser
 	}
 }
 
+function evaluateLazySchema(value: LazyItem<TreeNodeSchema>): TreeNodeSchema {
+	if (isLazy(value)) {
+		return (value as () => TreeNodeSchema)();
+	}
+	return value;
+}
+
 /**
  * @returns all types for which the data is schema-compatible.
  */
-export function getPossibleTypes(
-	globalSchema: FlexTreeSchema,
-	typeSet: AllowedTypeSet,
-	data: ContextuallyTypedNodeData,
-) {
-	// All types allowed by schema
-	const allowedTypes = getAllowedTypes(globalSchema, typeSet);
-
-	const possibleTypes: FlexTreeNodeSchema[] = [];
-	for (const allowed of allowedTypes) {
-		if (shallowCompatibilityTest(allowed, data)) {
-			possibleTypes.push(allowed);
+export function getPossibleTypes(typeSet: AllowedTypes, data: ContextuallyTypedNodeData) {
+	const possibleTypes: TreeNodeSchema[] = [];
+	for (const typeSetEntry of typeSet) {
+		const schema = evaluateLazySchema(typeSetEntry);
+		if (shallowCompatibilityTest(schema, data)) {
+			possibleTypes.push(schema);
 		}
 	}
 	return possibleTypes;
 }
 
+/**
+ * Normalized {@link ImplicitFieldSchema}.
+ */
+export interface NormalizedFieldSchema {
+	kind: FieldKind;
+	allowedTypes: AllowedTypes;
+}
+
+export function normalizeFieldSchema(schema: ImplicitFieldSchema): NormalizedFieldSchema {
```

Comment:
I think the approach used in FlexFieldSchema is better. Return a FieldSchema instance, and add the needed members to it to make that nice to use, mainly a lazily computed (so recursive cases work) set of allowed types.

Maybe generate and cache a Map<string,FieldSchema> (ex: weak map from simple-schema object to that) so you can have fast access to allowedTypes as a set of schema.

Mostly I'd like to make one nice APi for field schema instead of adding more types like NormalizedFieldSchema. SInce FieldSchema is a class we can cache derived data in it.

If you do keep NormalizedFieldSchema, make its fields readonly.

Reply:
Thanks. That lines up fairly well with my assumptions, just wanted to check before making further changes. I'll look at these options.

Reply:
> Maybe generate and cache a Map<string,FieldSchema> (ex: weak map from simple-schema object to that) so you can have fast access to allowedTypes as a set of schema.

How important do you think this is for performance here? I've applied your other suggestions, and the shape of the code looks much nicer. As far as I can tell, normalization here just incurs the allocation of a new FieldSchema object + (maybe) the allocation of an array when there is only a single allowed type.